File: tests/oracle/rule_addunsigned_typeprecond_1204.cc
/*
 * RULE-ADDUNSIGNED-TYPEPRECOND-0001
 *
 * Locked Ghidra 12.0.4 oracle for RuleAddUnsigned::getOpList/applyOp
 * (ruleaction.cc:7176-7214). The three cases isolate the read-facing
 * TYPE_UINT precondition from the high-quarter value test:
 *
 *   unknown_ff - newConstant's canonical TYPE_UNKNOWN + 0xff: reject
 *   uint_ff    - same-factory canonical TYPE_UINT + 0xff: rewrite to SUB 1
 *   uint_7f    - same-factory canonical TYPE_UINT + 0x7f: reject
 *
 * The projection records the rule return, opcode, operand/output identities,
 * replacement constant value/type/locks/symbol link, def-use bookkeeping,
 * alive/dead/block membership, and Varnode create indices. It deliberately
 * does not claim the op-aware union, character-print, real SymbolEntry/
 * EquateSymbol, or enum named-value branches.
 */

#include <bits/stdc++.h>

// Test-only access sets a bare namelock bit on uint_ff so copySymbol's exact
// typelock|namelock field transfer is observable without constructing a real
// EquateSymbol (which is an explicitly registered residual below).
#define private public
#include "architecture.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "ruleaction.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef private

#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::string;
using std::vector;

class FixtureTranslate final : public Translate {
  VarnodeData dummyRegister;

public:
  FixtureTranslate()
  {
    setBigEndian(false);
    setUniqueBase(0);
    insertSpace(new ConstantSpace(this,this));
    insertSpace(new AddrSpace(this,this,IPTR_PROCESSOR,"other",false,8,1,1,
                              AddrSpace::hasphysical,0,0));
    insertSpace(new UniqueSpace(this,this,2,0));
    AddrSpace *ram = new AddrSpace(this,this,IPTR_PROCESSOR,"ram",false,8,1,3,
                                   AddrSpace::hasphysical,0,0);
    insertSpace(ram);
    AddrSpace *reg = new AddrSpace(this,this,IPTR_PROCESSOR,"register",false,8,1,4,
                                   AddrSpace::hasphysical,0,0);
    insertSpace(reg);
    SpacebaseSpace *stack = new SpacebaseSpace(this,this,"stack",5,8,ram,1,true);
    insertSpace(stack);
    VarnodeData stackPointer;
    stackPointer.space = reg;
    stackPointer.offset = 0;
    stackPointer.size = 8;
    addSpacebasePointer(stack,stackPointer,8,true);
    insertSpace(new AddrSpace(this,this,IPTR_PROCESSOR,"join",false,8,1,6,
                              AddrSpace::hasphysical,0,0));
    insertSpace(new IopSpace(this,this,7));
    setDefaultCodeSpace(3);
    dummyRegister.space = reg;
    dummyRegister.offset = 0;
    dummyRegister.size = 8;
  }

  void initialize(DocumentStorage &) override {}
  const VarnodeData &getRegister(const string &) const override { return dummyRegister; }
  string getRegisterName(AddrSpace *,uintb,int4) const override { return ""; }
  string getExactRegisterName(AddrSpace *,uintb,int4) const override { return ""; }
  void getAllRegisters(std::map<VarnodeData,string> &) const override {}
  void getUserOpNames(vector<string> &) const override {}
  int4 instructionLength(const Address &) const override { return 0; }
  int4 oneInstruction(PcodeEmit &,const Address &) const override { return 0; }
  int4 printAssembly(AssemblyEmit &,const Address &) const override { return 0; }
};

class FixtureArchitecture final : public Architecture {
protected:
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}

public:
  FixtureArchitecture()
  {
    FixtureTranslate *fixtureTranslate = new FixtureTranslate();
    translate = fixtureTranslate;
    copySpaces(fixtureTranslate);
    max_basetype_size = 16;
    types = new TypeFactory(this);
    types->setupSizes();
    types->setCoreType("xunknown1",1,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown2",2,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown4",4,TYPE_UNKNOWN,false);
    types->setCoreType("xunknown8",8,TYPE_UNKNOWN,false);
    types->cacheCoreTypes();
    TypeOp::registerInstructions(inst,types,translate);
    symboltab = new Database(this,false);
    symboltab->attachScope(new ScopeInternal(0x101,"",this),(Scope *)0);
    ProtoModel *model = new ProtoModel(this);
    std::istringstream stream(
        "<prototype name=\"fixture\" extrapop=\"0\"><input/><output/></prototype>");
    XmlDecode decoder(this);
    decoder.ingestStream(stream);
    model->decode(decoder);
    protoModels[model->getName()] = model;
    setDefaultModel(model);
  }

  void printMessage(const string &) const override {}
};

static size_t countDescendants(const Varnode *vn)
{
  return static_cast<size_t>(std::distance(vn->beginDescend(),vn->endDescend()));
}

static const char *metaToken(type_metatype meta)
{
  switch(meta) {
  case TYPE_UNKNOWN: return "unknown";
  case TYPE_UINT: return "uint";
  default: return "other";
  }
}

static size_t countAlive(const Funcdata &fd)
{
  return static_cast<size_t>(std::distance(fd.beginOpAlive(),fd.endOpAlive()));
}

static size_t countDead(const Funcdata &fd)
{
  return static_cast<size_t>(std::distance(fd.beginOpDead(),fd.endOpDead()));
}

static bool aliveContains(const Funcdata &fd,const PcodeOp *op)
{
  for(list<PcodeOp *>::const_iterator iter=fd.beginOpAlive();iter!=fd.endOpAlive();++iter)
    if (*iter == op) return true;
  return false;
}

static size_t countBlockOps(const BlockBasic *block)
{
  return static_cast<size_t>(std::distance(block->beginOp(),block->endOp()));
}

static void emitState(const string &caseName,const string &stage,const string &result,
                      const Funcdata &fd,const BlockBasic *block,const PcodeOp *op,
                      const Varnode *input0,const Varnode *source,const Varnode *output,
                      const Datatype *uintType)
{
  const Varnode *current = op->getIn(1);
  const Datatype *currentType = current->getType();
  std::cout << "case=" << caseName
            << "|stage=" << stage
            << "|result=" << result
            << "|opcode=" << static_cast<int4>(op->code())
            << "|slot0_same=" << (op->getIn(0) == input0 ? 1 : 0)
            << "|slot1_is_source=" << (current == source ? 1 : 0)
            << "|new_constant=" << (current != source ? 1 : 0)
            << "|output_same=" << (op->getOut() == output ? 1 : 0)
            << "|value=" << current->getOffset()
            << "|size=" << current->getSize()
            << "|constant=" << (current->isConstant() ? 1 : 0)
            << "|type_meta=" << metaToken(currentType->getMetatype())
            << "|type_size=" << currentType->getSize()
            << "|type_same_source=" << (currentType == source->getType() ? 1 : 0)
            << "|type_is_factory_uint=" << (currentType == uintType ? 1 : 0)
            << "|type_lock=" << (current->isTypeLock() ? 1 : 0)
            << "|name_lock=" << (current->isNameLock() ? 1 : 0)
            << "|symbol_present=" << (current->getSymbolEntry() != (SymbolEntry *)0 ? 1 : 0)
            << "|symbol_same=" << (current->getSymbolEntry() == source->getSymbolEntry() ? 1 : 0)
            << "|source_desc=" << countDescendants(source)
            << "|current_desc=" << countDescendants(current)
            << "|input0_desc=" << countDescendants(input0)
            << "|output_def_is_op=" << (output->getDef() == op ? 1 : 0)
            << "|current_def_null=" << (current->getDef() == (PcodeOp *)0 ? 1 : 0)
            << "|alive_count=" << countAlive(fd)
            << "|alive_has_op=" << (aliveContains(fd,op) ? 1 : 0)
            << "|dead_count=" << countDead(fd)
            << "|block_count=" << countBlockOps(block)
            << "|parent_is_block=" << (op->getParent() == block ? 1 : 0)
            << "|op_dead=" << (op->isDead() ? 1 : 0)
            << "|source_create=" << source->getCreateIndex()
            << "|current_create=" << current->getCreateIndex()
            << "|input0_create=" << input0->getCreateIndex()
            << "|output_create=" << output->getCreateIndex()
            << '\n';
}

static void runCase(FixtureArchitecture &architecture,const string &caseName,
                    uintb value,bool installUint,bool lockTypeAndName)
{
  Scope *scope = architecture.symboltab->getGlobalScope();
  AddrSpace *ram = architecture.getSpaceByName("ram");
  AddrSpace *reg = architecture.getSpaceByName("register");
  if (scope == (Scope *)0 || ram == (AddrSpace *)0 || reg == (AddrSpace *)0)
    throw std::runtime_error("synthetic architecture is incomplete");

  Funcdata fd("fx","fx",scope,Address(ram,0x1000),(FunctionSymbol *)0,0x20);
  BlockGraph &graph = const_cast<BlockGraph &>(fd.getBasicBlocks());
  BlockBasic *block = graph.newBlockBasic(&fd);
  Varnode *input0 = fd.setInputVarnode(fd.newVarnode(1,reg,0x40));
  Varnode *source = fd.newConstant(1,value);
  Datatype *uintType = architecture.types->getBase(1,TYPE_UINT);
  if (installUint)
    source->updateType(uintType,lockTypeAndName,false);
  if (lockTypeAndName)
    source->setFlags(Varnode::namelock);

  PcodeOp *op = fd.newOp(2,Address(ram,0x2000));
  fd.opSetOpcode(op,CPUI_INT_ADD);
  Varnode *output = fd.newUniqueOut(1,op);
  fd.opSetInput(op,input0,0);
  fd.opSetInput(op,source,1);
  fd.opInsertEnd(op,block);

  emitState(caseName,"before","na",fd,block,op,input0,source,output,uintType);
  RuleAddUnsigned rule("analysis");
  int4 result = rule.applyOp(op,fd);
  emitState(caseName,"after",std::to_string(result),fd,block,op,input0,source,
            output,uintType);
}

static void run(void)
{
  vector<string> specPaths;
  startDecompilerLibrary(specPaths);
  {
    FixtureArchitecture architecture;
    RuleAddUnsigned rule("analysis");
    vector<uint4> opcodes;
    rule.getOpList(opcodes);
    std::cout << "getoplist:count=" << opcodes.size() << ",opcode="
              << (opcodes.empty() ? -1 : static_cast<int4>(opcodes[0])) << '\n';
    runCase(architecture,"unknown_ff",0xff,false,false);
    runCase(architecture,"uint_ff",0xff,true,true);
    runCase(architecture,"uint_7f",0x7f,true,false);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(void)
{
  try {
    run();
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/rule_addunsigned_typeprecond_1204.rs
//! Rust twin of `rule_addunsigned_typeprecond_1204.cc`.
//!
//! The covered projection invokes the production RuleAddUnsigned directly on
//! three equivalent IR graphs and emits the same record grammar as the locked
//! Ghidra 12.0.4 fixture. Factory-backed `TYPE_UNKNOWN` must reject, while a
//! canonical `TYPE_UINT` is required before the high-quarter value test.

use std::sync::{Arc, RwLock};

use rugra::action::Rule;
use rugra::address::Address;
use rugra::block::FlowBlock;
use rugra::funcdata::Funcdata;
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;
use rugra::ruleaction::RuleAddUnsigned;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeMetatype};
use rugra::type_system::typefactory::{CoreTypeFlavor, TypeFactory};
use rugra::varnode::{varnode_flags, Varnode};

type VarnodeRef = Arc<RwLock<Varnode>>;
type TypeFactoryRef = Arc<RwLock<TypeFactory>>;

fn meta_token(meta: TypeMetatype) -> &'static str {
    match meta {
        TypeMetatype::Unknown => "unknown",
        TypeMetatype::Uint => "uint",
        _ => "other",
    }
}

fn same_type(left: &Arc<Datatype>, right: &Arc<Datatype>) -> usize {
    usize::from(Arc::ptr_eq(left, right))
}

fn same_symbol(left: &Varnode, right: &Varnode) -> usize {
    match (&left.mapentry, &right.mapentry) {
        (Some(left), Some(right)) => usize::from(Arc::ptr_eq(left, right)),
        (None, None) => 1,
        _ => 0,
    }
}

#[allow(clippy::too_many_arguments)]
fn emit_state(
    case_name: &str,
    stage: &str,
    result: &str,
    fd: &Funcdata,
    block: &Arc<RwLock<dyn FlowBlock + Send + Sync>>,
    op: &PcodeOpRef,
    input0: &VarnodeRef,
    source: &VarnodeRef,
    output: &VarnodeRef,
    uint_type: &Arc<Datatype>,
) {
    let op_guard = op.0.read().unwrap();
    let current = op_guard.inrefs[1].clone();
    let current_guard = current.read().unwrap();
    let source_guard = source.read().unwrap();
    let input0_guard = input0.read().unwrap();
    let output_guard = output.read().unwrap();
    let current_type = current_guard
        .v_type
        .as_ref()
        .expect("factory-backed current type");
    let source_type = source_guard
        .v_type
        .as_ref()
        .expect("factory-backed source type");
    let output_def_is_op = output_guard
        .get_def()
        .is_some_and(|definition| Arc::ptr_eq(&definition, &op.0));
    let parent_is_block = op_guard
        .parent
        .as_ref()
        .and_then(std::sync::Weak::upgrade)
        .is_some_and(|parent| Arc::ptr_eq(&parent, block));
    let alive_has_op = fd
        .obank
        .alivelist
        .iter()
        .any(|candidate| Arc::ptr_eq(&candidate.0, &op.0));

    println!(
        "case={case_name}|stage={stage}|result={result}|opcode={}|slot0_same={}|slot1_is_source={}|new_constant={}|output_same={}|value={}|size={}|constant={}|type_meta={}|type_size={}|type_same_source={}|type_is_factory_uint={}|type_lock={}|name_lock={}|symbol_present={}|symbol_same={}|source_desc={}|current_desc={}|input0_desc={}|output_def_is_op={}|current_def_null={}|alive_count={}|alive_has_op={}|dead_count={}|block_count={}|parent_is_block={}|op_dead={}|source_create={}|current_create={}|input0_create={}|output_create={}",
        op_guard.opcode as i32,
        usize::from(Arc::ptr_eq(&op_guard.inrefs[0], input0)),
        usize::from(Arc::ptr_eq(&current, source)),
        usize::from(!Arc::ptr_eq(&current, source)),
        usize::from(op_guard.output.as_ref().is_some_and(|candidate| Arc::ptr_eq(candidate, output))),
        current_guard.get_offset(),
        current_guard.get_size(),
        usize::from(current_guard.is_constant()),
        meta_token(current_type.get_metatype()),
        current_type.get_size(),
        same_type(current_type, source_type),
        same_type(current_type, uint_type),
        usize::from(current_guard.is_type_lock()),
        usize::from(current_guard.is_name_lock()),
        usize::from(current_guard.mapentry.is_some()),
        same_symbol(&current_guard, &source_guard),
        source_guard.count_descends(),
        current_guard.count_descends(),
        input0_guard.count_descends(),
        usize::from(output_def_is_op),
        usize::from(current_guard.get_def().is_none()),
        fd.obank.alivelist.len(),
        usize::from(alive_has_op),
        fd.obank.deadlist.len(),
        block.read().unwrap().get_ops().len(),
        usize::from(parent_is_block),
        usize::from(op_guard.is_dead()),
        source_guard.get_create_index(),
        current_guard.get_create_index(),
        input0_guard.get_create_index(),
        output_guard.get_create_index(),
    );
}

fn run_case(
    type_factory: &TypeFactoryRef,
    case_name: &str,
    value: u64,
    install_uint: bool,
    lock_type_and_name: bool,
) {
    let mut fd = Funcdata::new("fx", Address::new(0x1000), 0x20);
    fd.vbank.set_type_factory(type_factory.clone());
    let block = fd.create_new_block();
    let input0 = fd.vbank.create_with_space(1, AddressSpace::Register, 0x40);
    let input0 = fd.set_input_varnode(input0);
    let source = fd.new_constant(1, value);
    let uint_type = type_factory
        .read()
        .unwrap()
        .get_base(1, TypeMetatype::Uint)
        .expect("canonical uint1");
    if install_uint {
        source
            .write()
            .unwrap()
            .update_type_lock(uint_type.clone(), lock_type_and_name, false);
    }
    if lock_type_and_name {
        source.write().unwrap().set_flags(varnode_flags::NAMELOCK);
    }

    let op = fd.new_op(2, Address::new(0x2000));
    fd.op_set_opcode(&op, OpCode::CPUI_INT_ADD);
    let output = fd.new_unique_out(1, &op);
    fd.op_set_input(&op, input0.clone(), 0);
    fd.op_set_input(&op, source.clone(), 1);
    fd.op_insert_end(&op, &block);

    emit_state(
        case_name, "before", "na", &fd, &block, &op, &input0, &source, &output, &uint_type,
    );
    let result = RuleAddUnsigned::new()
        .apply_op(&op.0, &mut fd)
        .expect("RuleAddUnsigned apply");
    emit_state(
        case_name,
        "after",
        &result.to_string(),
        &fd,
        &block,
        &op,
        &input0,
        &source,
        &output,
        &uint_type,
    );
}

fn main() {
    let opcodes = RuleAddUnsigned::new().get_opcodes();
    println!(
        "getoplist:count={},opcode={}",
        opcodes.len(),
        opcodes.first().map_or(-1, |opcode| *opcode as i32),
    );

    // One synthetic Architecture-equivalent factory is shared across all
    // cases. Both source and replacement constants therefore resolve inside
    // the same canonical identity domain, matching `glb->types` in Ghidra.
    let type_factory = Arc::new(RwLock::new(TypeFactory::new_flavor(
        8,
        CoreTypeFlavor::Standalone,
    )));
    run_case(&type_factory, "unknown_ff", 0xff, false, false);
    run_case(&type_factory, "uint_ff", 0xff, true, true);
    run_case(&type_factory, "uint_7f", 0x7f, true, false);
}
